rust/chains/abacus-ethereum/src/mailbox.rs

```diff
@@ -0,0 +1,318 @@
+#![allow(clippy::enum_variant_names)]
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::{error::Error as StdError, sync::Arc};
+
+use async_trait::async_trait;
+use ethers::abi::AbiEncode;
+use ethers::prelude::*;
+use ethers_contract::builders::ContractCall;
+use eyre::{eyre, Result};
+use tracing::instrument;
+
+use abacus_core::{
+    AbacusAbi, AbacusContract, AbacusMessage, ChainCommunicationError, Checkpoint, ContractLocator,
+    Indexer, LogMeta, Mailbox, MailboxIndexer, RawAbacusMessage, TxCostEstimate, TxOutcome,
+};
+
+use crate::contracts::mailbox::{Mailbox as EthereumMailboxInternal, ProcessCall, MAILBOX_ABI};
+use crate::trait_builder::MakeableWithProvider;
+use crate::tx::report_tx;
+
+impl<M> std::fmt::Display for EthereumMailboxInternal<M>
+where
+    M: Middleware,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub struct MailboxIndexerBuilder {
+    pub finality_blocks: u32,
+}
+
+impl MakeableWithProvider for MailboxIndexerBuilder {
+    type Output = Box<dyn MailboxIndexer>;
+
+    fn make_with_provider<M: Middleware + 'static>(
+        &self,
+        provider: M,
+        locator: &ContractLocator,
+    ) -> Self::Output {
+        Box::new(EthereumMailboxIndexer::new(
+            Arc::new(provider),
+            locator,
+            self.finality_blocks,
+        ))
+    }
+}
+
+#[derive(Debug)]
+/// Struct that retrieves event data for an Ethereum mailbox
+pub struct EthereumMailboxIndexer<M>
+where
+    M: Middleware,
+{
+    contract: Arc<EthereumMailboxInternal<M>>,
+    provider: Arc<M>,
+    finality_blocks: u32,
+}
+
+impl<M> EthereumMailboxIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    /// Create new EthereumMailboxIndexer
+    pub fn new(provider: Arc<M>, locator: &ContractLocator, finality_blocks: u32) -> Self {
+        let contract = Arc::new(EthereumMailboxInternal::new(
+            &locator.address,
+            provider.clone(),
+        ));
+        Self {
+            contract,
+            provider,
+            finality_blocks,
+        }
+    }
+}
+
+#[async_trait]
+impl<M> Indexer for EthereumMailboxIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    #[instrument(err, skip(self))]
+    async fn get_finalized_block_number(&self) -> Result<u32> {
+        Ok(self
+            .provider
+            .get_block_number()
+            .await?
+            .as_u32()
+            .saturating_sub(self.finality_blocks))
+    }
+}
+
+#[async_trait]
+impl<M> MailboxIndexer for EthereumMailboxIndexer<M>
+where
+    M: Middleware + 'static,
+{
+    #[instrument(err, skip(self))]
+    async fn fetch_sorted_messages(
+        &self,
+        from: u32,
+        to: u32,
+    ) -> Result<Vec<(RawAbacusMessage, LogMeta)>> {
```

Comment:
I wonder if we should just have this return `Result<Vec<(AbacusMessage, LogMeta)>>` so that we can also avoid a bunch of `AbacusMessage::from` creations when sorting